### PR TITLE
Checks

### PR DESCRIPTION
**Pull request by orhanuysal-google**

### CODE REVIEW

File: src/checks.cc [hunk 1]
```diff
@@ -349,10 +349,11 @@ LinterResult CheckJoin(absl::string_view sql, const LinterOptions &option) {
            LinterResult result;
            // SingleNodeDebugString also returns the type if there are any.
            // If it is equal to normal kind string, this means join is typeless.
+           int position = node->GetParseLocationRange().start().GetByteOffset();
```
Comment (a-litvinov) [inline src/checks.cc:353]: Nit: you're fetching a node location in few places. If it happens more than twice, you could write a tiny inlined helper functions `int GetStartPosition(const ASTNode&)`, `GetEndPosition` and the same for `const Token&`

File: docs/checks.md [hunk 2]
```diff
@@ -303,5 +303,57 @@ In line 3, column 14: PROTO and MODULE inputs should be in seperate groups. [imp
...
+SELECT COUNT(*); -- line 3
+```
 
+**Linter Output**
+```
+In line 2, column 15: Use COUNT(*) instead of COUNT(1) [count-star]
+```
+
+## keyword-identifier
+If the name of an identifier(column name, parameter name etc.) is also a SQL keyword. Change the name or use escape characters(`).
```
Comment (a-litvinov) [inline docs/checks.md:342]: Do not use ZetaSQL keywords as identifiers (aliases, function arguments, etc.). Change the name or escape with backticks (`).

Btw, I don't remember what exactly you're checking there. I thought you've limited this check to aliases only.

File: docs/checks.md [hunk 3]
```diff
@@ -303,5 +303,57 @@ In line 3, column 14: PROTO and MODULE inputs should be in seperate groups. [imp
...
+SELECT MyTable.travel_date AS t_date; -- line 7
+
+-- Escape characters can also be used.
+SELECT type FROM A;
+SELECT `type` FROM A; -- line 7
+```
+
+**Linter Output**
+```
+In line 2, column 31: Identifier `Date` is an SQL keyword. Use escape charater(`) to fix it [keyword-identifier]
```
Comment (a-litvinov) [inline docs/checks.md:357]: Actually, can you also update the error message:

Identifier <> is an SQL keyword. Change the name or escape with backticks (`)

File: src/checks.cc [hunk 4]
```diff
@@ -403,8 +404,33 @@ LinterResult CheckImports(absl::string_view sql, const LinterOptions &option) {
...
+               int start = pos - 1;
+               int end = node->GetParseLocationRange().end().GetByteOffset();
+
+               if (IgnoreBackwardSpaces(sql, &start) ||
+                   IgnoreForwardSpaces(sql, &end) || sql[start] != '(' ||
+                   sql[end] != ')')
+                 if (option.IsActive(ErrorCode::kExpressionParanteses, pos))
+                   result.Add(ErrorCode::kExpressionParanteses, sql, pos,
+                              "Use parantheses between consequtive AND and OR "
+                              "statements.");
```
Comment (a-litvinov) [inline src/checks.cc:428]: s/statements/operators/

File: src/checks.cc [hunk 5]
```diff
@@ -436,7 +462,25 @@ LinterResult CheckCountStar(absl::string_view sql,
 LinterResult CheckKeywordNamedIdentifier(absl::string_view sql,
                                          const LinterOptions &option) {
   LinterResult result;
+  std::vector<ParseToken> keywords = GetKeywords(sql);
```
Comment (a-litvinov) [inline src/checks.cc:465]: If you choose my suggestion to extract file parsing into separate function, the second step would be to extract all keywords and identifiers also only once, before all checks :)

File: src/checks_test.cc [hunk 6]
```diff
@@ -306,5 +306,52 @@ TEST(LinterTest, CheckImports) {
...
+      CheckExpressionParantheses("SELECT D where a and b or c;", option).ok());
+  EXPECT_FALSE(
+      CheckExpressionParantheses("SELECT D where a or b and c;", option).ok());
+  EXPECT_TRUE(CheckExpressionParantheses(
+                  "SELECT D where (       a or b   ) and c;", option)
+                  .ok());
+}
+
+TEST(LinterTest, CheckCountStar) {
+  LinterOptions option;
```
Comment (a-litvinov) [inline src/checks_test.cc:333]: Add few more tests here:
```
SELECT count ( 1 )
/* count(1) */ SELECT COUNT(*) -- count 1
```

File: src/checks_test.cc [hunk 7]
```diff
@@ -306,5 +306,52 @@ TEST(LinterTest, CheckImports) {
...
+  EXPECT_TRUE(CheckCountStar("SELECT COUNT", option).ok());
+  EXPECT_TRUE(CheckCountStar("SELECT COUNT(0)", option).ok());
+  EXPECT_FALSE(CheckCountStar("SELECT COUNT(1)", option).ok());
+  EXPECT_TRUE(CheckCountStar("SELECT COUNT(*)", option).ok());
+}
+
+TEST(LinterTest, CheckKeywordNamedIdentifier) {
+  LinterOptions option;
+  LinterResult result = CheckKeywordNamedIdentifier("SELECT Date", option);
+  result.PrintResult();
```
Comment (a-litvinov) [inline src/checks_test.cc:343]: nit: good practice is to make tests print stuff only if there is an error. This way, you can quickly see in the test output the error log from the failing test.

File: src/checks_util.cc [hunk 8]
```diff
@@ -37,11 +37,16 @@
 
 namespace zetasql::linter {
 
+absl::string_view GetName(const ParseLocationRange &range,
+                          const absl::string_view &sql) {
+  const int &start = range.start().GetByteOffset();
```
Comment (a-litvinov) [inline src/checks_util.cc:42]: `const int start` and same for `end`

File: src/checks.cc [hunk 9]
```diff
@@ -116,7 +116,8 @@ LinterResult CheckUppercaseKeywords(absl::string_view sql,
       if (option.IsActive(ErrorCode::kLetterCase, position))
         result.Add(
             ErrorCode::kLetterCase, sql, position,
-            absl::StrCat("All keywords should be ",
+            absl::StrCat("Keyword '", GetName(token.GetLocationRange(), sql),
```
Comment (a-litvinov) [inline src/checks.cc:119]: Why you don't use `token.GetImage()` ? Here and in one more place. Then you could keep GetName for ASTNode only. (I wonder, if ASTNode also has a similar getter).

**Review by a-litvinov — CHANGES_REQUESTED** [review]

File: src/linter.cc [hunk 10]
```diff
@@ -45,7 +45,8 @@ LinterResult ParseNoLintSingleComment(absl::string_view line,
   LinterResult result;
   std::map<std::string, ErrorCode> error_map = GetErrorMap();
 
-  const LazyRE2 kRegex = {"\\s*(NOLINT|LINT)\\s*\\(([a-z ,-]*)\\)\\s*(.*)\\s*"};
+  const LazyRE2 kRegex = {
+      "\\s*(NOLINT|LINT)\\s*\\(([a-z ,\"-]*)\\)\\s*(.*)\\s*"};
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:52]: On Friday, I wrongly type nolint with quotes and it wasn't working. I tried to debug. Then figured out. So, if i include quotes in there nothing will change except there will be specific nolint usage error for the ones that used quote.
 
Oh, I didn't know about about the raw string literal.

[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/linter.cc:52]: Can you paste an example here? What kind of comment you're trying to catch with this regexp?
Also, I'm wondering why do you catch both NOLINT and LINT? Do you allow disabling lint checks by saying -- LINT(check-name) ?

[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:52]: For example If user say:
`// NOLINT("join")`
Error Message will be:
`In line 1, column 18: Unkown NOLINT error category: '"join"' [nolint]`
becuase they added extra quotation. Also, I mentioned in documentation, LINT will re-enable check, while NOLINT will disable it. Like:
```
-- NOLINT(join)
/* ... some special code */
-- LINT(join)
-- Here join check is active.
```

File: src/linter.cc [hunk 11]
```diff
@@ -122,20 +124,19 @@ LinterOptions GetOptionsFromConfig(Config config, absl::string_view filename) {
 
   std::map<std::string, ErrorCode> error_map = GetErrorMap();
 
-  for (std::string check_name : config.nolint()) {
-    if (error_map.count(check_name)) {
+  for (std::string check_name : config.nolint())
+    if (error_map.count(check_name))
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:157]: Fixed

File: src/linter.cc [hunk 12]
```diff
@@ -122,20 +124,19 @@ LinterOptions GetOptionsFromConfig(Config config, absl::string_view filename) {
...
       option.DisactivateCheck(error_map[check_name]);
-    }
-  }
+
   return option;
 }
 
-LinterResult RunChecks(absl::string_view sql, LinterOptions options) {
+LinterResult RunChecks(absl::string_view sql, LinterOptions option) {
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:163]: Actually, I changed them everywhere :p

File: src/runner.cc [hunk 13]
```diff
@@ -38,6 +38,9 @@ ABSL_FLAG(bool, quick, false,
           "This flag will read from standart input. It will read one"
           "statement and continue until reading semicolon ';'");
 
+ABSL_FLAG(bool, debug, false,
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/runner.cc:41]: changed it as your example.

File: src/linter.cc [hunk 14]
```diff
@@ -106,6 +109,32 @@ LinterResult ParseNoLintComments(absl::string_view sql,
   return result;
 }
 
+LinterResult CheckParserSucceeds(absl::string_view sql, LinterOptions* option) {
+  ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
+  bool is_the_end = false;
+  int byte_position = 1;
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:115]: Sorry, It was not necessary anymore, (I get byte_position directly from status.) So erased it.

File: src/linter.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/linter.cc:52]: Ah, thanks for the clarification! So you want to be flexible and allow both `NOLINT(join)` and `NOLINT("join")`?

File: src/linter.cc [hunk 15]
```diff
@@ -106,6 +109,32 @@ LinterResult ParseNoLintComments(absl::string_view sql,
...
+                                                   &output, &is_the_end);
+    if (!status.ok()) {
+      if (option->IsActive(ErrorCode::kParseFailed, byte_position)) {
+        ErrorLocation position = internal::GetPayload<ErrorLocation>(status);
+        result.Add(ErrorCode::kParseFailed, position.line(), position.column(),
+                   status.message());
+        return result;
+      }
+    }
+    option->AddParserOutput(output.release());
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:131]: I tried it, and spent hours to debug etc.  It seems you can't create `vector<std::unique_ptr<AnyType>>` as long as you have instances of the class. So, I choose this way, I will try to make it unique_ptr again though.

File: src/linter_options.h [hunk 16]
```diff
@@ -63,45 +66,55 @@ class LinterOptions {
   // INCREASING order.
   void Enable(ErrorCode code, int position);
 
+  // Adds a single parser output to parset_output_
+  void AddParserOutput(ParserOutput *output);
+
+  // ---------------------------------- GETTER/SETTER functions
+
+  // Getter for parser_output_
+  std::vector<ParserOutput *> ParserOutputs() const { return parser_outputs_; }
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter_options.h:77]: Yes.

File: src/linter_options.h [hunk 17]
```diff
@@ -134,6 +147,13 @@ class LinterOptions {
   // options for that check.
   std::map<ErrorCode, CheckOptions> option_map_;
 
+  // Stores Whether parser output should be saved.
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter_options.h:152]: Yes, I will change its explanation. I think it is needed though. Parser may be invoked multiple times if there are some errors that  makes parser_outputs_ empty.

File: src/linter_options.h [hunk 18]
```diff
@@ -63,45 +66,55 @@ class LinterOptions {
   // INCREASING order.
   void Enable(ErrorCode code, int position);
 
+  // Adds a single parser output to parset_output_
+  void AddParserOutput(ParserOutput *output);
+
+  // ---------------------------------- GETTER/SETTER functions
+
+  // Getter for parser_output_
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter_options.h:76]: Erased extra comments. I also think it would be better without comments, but there is a rule that every function should have an explanation comment, So, one approval to erase these comments is all I need :)

File: src/checks_util.cc [hunk 19]
```diff
@@ -252,4 +266,49 @@ zetasql_base::StatusOr<VisitResult> RuleVisitor::defaultVisit(
...
 
+std::vector<ParseToken> GetKeywords(absl::string_view sql) {
+  ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
+  std::vector<ParseToken> parse_tokens;
+  std::vector<ParseToken> keywords;
+
+  absl::Status status =
+      GetParseTokens(ParseTokenOptions(), &location, &parse_tokens);
+
+  if (!status.ok()) return keywords;
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:296]: I thought if tokenizer doesn't work, there should be some errror message in some other tool or the sql is good as it is in a different dialect. In both case printing an error is not linter's job to do (preventing multiple errors). Also this function called several times(again preventing multiple errors).

Returning empty string simply makes the check return without errors, and I think it is intended behavior when tokenizer fails, or is it?

File: src/checks_util.cc [hunk 20]
```diff
@@ -252,4 +266,49 @@ zetasql_base::StatusOr<VisitResult> RuleVisitor::defaultVisit(
...
+  for (auto &token : parse_tokens)
+    if (token.kind() == ParseToken::KEYWORD) keywords.push_back(token);
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:298]: Yes, there were really lots of normal identifiers that tokenizer wasn't sure and identified them as Keyword_Or_Identifier. I didn't want to include them.

File: src/linter_test.cc [hunk 21]
```diff
@@ -62,5 +62,25 @@ TEST(NolintTest, MultipleDisabling) {
   EXPECT_TRUE(errors[3].GetPosition() == std::make_pair(8, 2));
 }
 
+TEST(LinterTest, StatementValidityCheck) {
+  LinterOptions option;
+  EXPECT_TRUE(CheckParserSucceeds("SELECT 5+2", &option).ok());
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter_test.cc:67]: Yes, it accumulates, but that will never be used. I added a comment for this behavior.

For another thought: CheckParserSucceeds just adds an optimization flag and AST to options. So even when if it wasn't invoked, linter will be normal but slower since multiple parser calls will be made. I currently made "RunChecks" so that CheckParserSucceeds will run before all checks.

File: src/checks_util.cc [hunk 22]
```diff
@@ -252,4 +266,49 @@ zetasql_base::StatusOr<VisitResult> RuleVisitor::defaultVisit(
...
+
+  sort(keywords.begin(), keywords.end(),
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:300]: I did just to be sure, but ok. I am erasing it...

File: src/checks_util.cc [hunk 23]
```diff
@@ -88,6 +90,18 @@ bool IsLowerSnakeCase(absl::string_view name) {
   return true;
 }
 
+bool IsBefore(const ASTNode *node, const ParseToken &token) {
+  return node->GetParseLocationRange().start().GetByteOffset() <
+         token.GetLocationRange().start().GetByteOffset();
+}
+
+bool IsTheSame(const ASTNode *node, const ParseToken &token) {
+  return node->GetParseLocationRange().start().GetByteOffset() ==
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:104]: Oh, I see '==' operator is defined for  ParseLocationRange, fixed it.

File: src/checks_util.h [hunk 24]
```diff
@@ -141,6 +149,18 @@ bool ConsistentUppercaseLowercase(const absl::string_view &sql,
                                   const ParseLocationRange &range,
                                   const LinterOptions &option);
 
+// Return all tokenizer keywords of a sql query.
+// They will be returned sorted in ascending position order.
+std::vector<ParseToken> GetKeywords(absl::string_view sql);
+
+// Helper function that adds all identifiers in subtree of a ASTNode
+// to a list.
+void GetIdentifiers(const ASTNode *node, std::vector<const ASTNode *> *list);
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.h:171]: I will test it, and it is a general function that gives identifiers of AST. It could potentially be used.

File: src/checks_util.h [hunk 25]
```diff
@@ -116,6 +116,16 @@ bool IsBefore(const ASTNode *node, const ParseToken &token);
...
 
+// Ignores forward spaces and increase the position until the first non-empty
+// character. (empty characters: ' ', '\t', '\n')
+// Returns true if position reaches to the end.
+bool IgnoreForwardSpaces(absl::string_view sql, int *position);
+
+// Ignores backward spaces and decrease the position until the first non-empty
+// character. (empty characters: ' ', '\t', '\n')
+// Returns true if position reaches to the end.
+bool IgnoreBackwardSpaces(absl::string_view sql, int *position);
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.h:130]: Ok :)

File: src/checks.cc [hunk 26]
```diff
@@ -408,7 +408,26 @@ LinterResult CheckExpressionParantheses(absl::string_view sql,
 LinterResult CheckCountStar(absl::string_view sql,
                             const LinterOptions &option) {
   LinterResult result;
-
+  for (int i = 0; i < static_cast<int>(sql.size()); i++) {
+    IgnoreComments(sql, option, &i);
+    IgnoreStrings(sql, &i);
+    if (i + 5 < static_cast<int>(sql.size()) &&
+        ConvertToUppercase(sql.substr(i, 5)) == "COUNT") {
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks.cc:443]: Sure, I uesd it.

File: docs/checks.md (hunk 2 above)
Comment (orhanuysal-google) [reply docs/checks.md:342]: It was another update that makes the check "column-name" aliases only. This is a seperate check that warns keyword-named identifiers.

File: docs/checks.md (hunk 3 above)
Comment (orhanuysal-google) [reply docs/checks.md:357]: Sure

File: src/checks.cc (hunk 4 above)
Comment (orhanuysal-google) [reply src/checks.cc:428]: Made it "operators" (and there shouldn't be '.' at the end, since there will be '[<>]' )

File: src/checks.cc (hunk 5 above)
Comment (orhanuysal-google) [reply src/checks.cc:465]: I think "RunChecks" is a seperate function for adding sth. before all checks. So, I can add one line to it to save keywords/Tokens etc. But, I don't think it would be effective. Most checks needs to traverse AST tree to apply specific functions and just traversing is not costly at all compared to whole parser. As for tokenizer, There is literally 2 checks that using it.

So, I think the can be added to options to optimize by creating these only once, but it doesn't change much (20 checks having ~O(4N), we will have ~O(N/10) identifiers that will be created twice, I could say roughly <1%). I am leaving it to future.

File: src/checks_test.cc (hunk 7 above)
Comment (orhanuysal-google) [reply src/checks_test.cc:343]: Yes, I erased it. (forgot that in there.)

File: src/checks_test.cc (hunk 6 above)
Comment (orhanuysal-google) [reply src/checks_test.cc:333]: I added, along with inside string test.

File: src/checks_util.cc (hunk 8 above)
Comment (orhanuysal-google) [reply src/checks_util.cc:42]: Ah yes, its just integer, so pointer actually more costly. (4 vs. 8 byte)

File: src/checks.cc [hunk 27]
```diff
@@ -95,40 +96,28 @@ LinterResult CheckSemicolon(absl::string_view sql,
...
-        for (auto word : nolint_words)
-          if (token.GetSQL() == word) nolint = true;
-        if (nolint) continue;
-        if (option.IsActive(ErrorCode::kLetterCase, position))
-          result.Add(
-              ErrorCode::kLetterCase, sql, position,
-              absl::StrCat("All keywords should be ",
-                           option.UpperKeyword() ? "uppercase" : "lowercase"));
-      }
+    if (token.GetSQL() == "TRUE" || token.GetSQL() == "FALSE") continue;
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks.cc:112]: Were true and false keywords that should be capitalized? I understand they are keywords now (I thought they weren't) I am erasing it.

File: src/checks_util.h [hunk 28]
```diff
@@ -141,6 +149,18 @@ bool ConsistentUppercaseLowercase(const absl::string_view &sql,
                                   const ParseLocationRange &range,
                                   const LinterOptions &option);
 
+// Return all tokenizer keywords of a sql query.
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.h:165]: ok

File: src/checks.cc (hunk 9 above)
Comment (orhanuysal-google) [reply src/checks.cc:119]: I don't know why but I remember it was automaticly capitalized in tokenizer. Now I see thats not true in everycase. So changed it to `token.GetImage()`

File: src/checks_util.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/checks_util.cc:293]: well, we should let the user know that some lint checks were skipped for their file even though they didn't enable them. The error could instruct the user to disable this lint check if it is some valid code, but tokenizer fails.

File: src/linter.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc:52]: Actually, force them to say `NOLINT(join)` by including an error message for 'NOLINT("join")'

File: src/runner.cc [hunk 29]
```diff
@@ -128,7 +127,18 @@ void run(std::vector<std::string> sql_files, Config config) {
 }  // namespace
 }  // namespace zetasql::linter
 
+class B {
```
Comment (a-litvinov) [inline src/runner.cc:130]: Accidental debug stuff?

File: src/checks.cc (hunk 5 above)
Comment (a-litvinov) [reply src/checks.cc:464]: sounds good

File: src/checks_util.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:293]: Then I will have a tokenizer check that will say "tokenizer failed, and [this, this and this] checks will not work." etc.

File: src/runner.cc (hunk 29 above)
Comment (orhanuysal-google) [reply src/runner.cc]: Yes, it was for testing `std::vector<std::unique_pointer<bool>> non_used_variable;` when I added this to LinterOptions I got errors, and I think I can't convert to it.

I erased it, but then it returned when I try to recover stuff with ctrl-z (didn't realized that)

Comment (orhanuysal-google) [reply src/runner.cc]: it is fixed now.

File: src/linter.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/linter.cc]: I assume your problem was that you were doing copies of LinterOptions, and having unique_ptr members disables default copy and move constructors. Can you try again?

File: src/linter.cc [hunk 30]
```diff
@@ -104,36 +109,69 @@ LinterResult ParseNoLintComments(absl::string_view sql,
...
   if (config.has_upper_keyword())
-    option.SetUpperKeyword(config.upper_keyword());
+    options.SetUpperKeyword(config.upper_keyword());
 
   std::map<std::string, ErrorCode> error_map = GetErrorMap();
 
   for (std::string check_name : config.nolint()) {
     if (error_map.count(check_name)) {
-      option.DisactivateCheck(error_map[check_name]);
+      options.DisactivateCheck(error_map[check_name]);
```
Comment (a-litvinov) [inline src/linter.cc:159]: nit: "disactivate" is almost never used in English. The used alternative is "deactivate", but in this context I'd say "DisableCheck"

File: src/runner.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/runner.cc]: print_parsed_ast or print_ast would be more self-explanatory name for the flag.

File: src/linter_test.cc (hunk 21 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/linter_test.cc]: What happens if the user disables parser-succeeds check? Can you extract file parsing in a completely separate function called before all checks? It could store the parser output or the error if there was a parsing error. Then if CheckParserSucceeds is enabled, it would just check if you have stored parser output and no parsing errors. Also, this will make the order of checks irrelevant.

File: src/checks_util.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/checks_util.cc:293]: I think this might be error-prone. What if someone adds a new check that depends on tokenizer? they will have to know that they need also find a tokenizer check and see what error message it prints. I'm sure I'd forget this myself.
I think, you could print a warning message directly in each individual check saying that the check is skipped since the tokenizer failed.
Or make it so, that when registering a new check one must specify what dependencies it has. Then there will be a registry of checks that depend on parser or tokenizer.

File: src/checks_util.h (hunk 24 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/checks_util.h:171]: sounds good

File: docs/checks.md [hunk 31]
```diff
@@ -303,5 +303,57 @@ In line 3, column 14: PROTO and MODULE inputs should be in seperate groups. [imp
...
+-- Here is an invalid usage of operators.
+SELECT D WHERE a AND b OR c; -- line 5
+
+-- It can be fixed by adding paratheses.
+SELECT D WHERE a AND (b OR c);
+```
+
+**Linter Output**
+```
+In line 5, column 16: Use parantheses between consequtive AND and OR statements. [expression-parantheses]
```
Comment (a-litvinov) [inline docs/checks.md:323]: Update the error message, since you've updated it in the code

File: src/checks_test.cc (hunk 6 above)
Comment (a-litvinov) [reply src/checks_test.cc]: I meant two separate EXPECT_* :D

File: src/checks.cc [hunk 32]
```diff
@@ -436,8 +402,91 @@ LinterResult CheckImports(absl::string_view sql, const LinterOptions &option) {
...
+    while (index < identifiers.size() && IsBefore(identifiers[index], token))
+      index++;
+
+    // The Identifier is also a keyword
+    if (index < identifiers.size() && IsTheSame(identifiers[index], token)) {
+      int position = token.GetLocationRange().start().GetByteOffset();
+      if (options.IsActive(ErrorCode::kKeywordIdentifier, position))
+        result.Add(
+            ErrorCode::kKeywordIdentifier, sql, position,
+            absl::StrCat("Identifier `", GetName(token.GetLocationRange(), sql),
```
Comment (a-litvinov) [inline src/checks.cc:478]: token.GetImage()

File: src/checks_util.cc [hunk 33]
```diff
@@ -32,14 +33,20 @@
 #include "zetasql/public/parse_helpers.h"
 #include "zetasql/public/parse_location.h"
 #include "zetasql/public/parse_resume_location.h"
+#include "zetasql/public/parse_tokens.h"
 
 namespace zetasql::linter {
 
+absl::string_view GetName(const ParseLocationRange &range,
```
Comment (a-litvinov) [inline src/checks_util.cc:40]: Since you can use token.GetImage() for tokens, you can change this function to `absl::string_view GetName(const ASTNode& node)`

File: src/runner.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/runner.cc]: Oh, my bad :s

File: src/linter.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter.cc]: Took a while to find all copy operations (Error message gives location in std:: library), but i finally made it unique_ptr.

File: src/linter_test.cc (hunk 21 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/linter_test.cc]: "CheckParserSucceeds" is completely seperate function that doesn't count as a "check" since does not follow the signature and of course it can't be in checks_list. This function will store parser output, or return an error if there was a parsing error. It can also be enabled/disabled. If it is enabled and parser succeeds, Parser-Dependant checks will work fine, otherwise Parser-Dependant Checks will also be skipped.

I think clang-tidy/cpplint only prints actual fixable lint errors. So, there shouldn't be skipped message for each check when parser is disabled or failed. It should just ignore, which is the current implementation.

File: src/checks_util.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks_util.cc:293]: "I think, you could print a warning message directly in each individual check saying that the check is skipped since the tokenizer failed."
Ok, I am adding it.

File: src/checks.cc (hunk 1 above)
Comment (orhanuysal-google) [reply src/checks.cc:352]: Ok.

File: src/runner.cc (hunk 29 above)
Comment (a-litvinov) [reply src/runner.cc:130]: ahm, you brought it back in the last commit :D

File: src/checks_test.cc (hunk 6 above)
Comment (orhanuysal-google) [reply src/checks_test.cc]: Oh.. anyway, It is stronger this way :D
Also added seperate for smaller one.

File: src/linter.cc (hunk 30 above)
Comment (orhanuysal-google) [reply src/linter.cc]: fixed

File: docs/checks.md (hunk 31 above)
Comment (orhanuysal-google) [reply docs/checks.md:323]: fixed

File: src/checks.cc (hunk 32 above)
Comment (orhanuysal-google) [reply src/checks.cc:478]: Fixed

File: src/checks_util.cc (hunk 33 above)
Comment (orhanuysal-google) [reply src/checks_util.cc:40]: There is already GetNodeString for that

File: src/runner.cc (hunk 29 above)
Comment (orhanuysal-google) [reply src/runner.cc]: Oh, it was back at 'unique_ptr' commit, but erased again in the last one 'review changes'. So, it is fine now.

File: src/linter_test.cc (hunk 21 above)
[the remark this answers is not in the document]
Comment (a-litvinov) [reply src/linter_test.cc]: ah, ok, sounds good!

File: src/checks_util.cc [hunk 34]
```diff
@@ -282,15 +289,19 @@ zetasql_base::StatusOr<VisitResult> RuleVisitor::defaultVisit(
   return VisitResult::VisitChildren(node);
 }
 
-std::vector<ParseToken> GetKeywords(absl::string_view sql) {
+std::vector<ParseToken> GetKeywords(absl::string_view sql, ErrorCode code) {
```
Comment (a-litvinov) [inline src/checks_util.cc:292]: Oh, this ErrorCode is really confusing here. Should we rename it to LintCheckCode or something? (Not in this pull request, please :)

Comment (orhanuysal-google) [reply src/checks_util.cc:292]: Hmm, it is ErrorCode inside LintError, but it also correspond to check (each check has different error code.) So, yes, we can rename it, but it is a very core class.

Comment (a-litvinov) [reply src/checks_util.cc:292]: Yes, I understand what it means, but I scratched my head for 30 seconds, before I figured this out :) Let's rename in a separate cleanup PR.

File: src/checks_util.cc [hunk 35]
```diff
@@ -282,15 +289,19 @@ zetasql_base::StatusOr<VisitResult> RuleVisitor::defaultVisit(
...
   ParseResumeLocation location = ParseResumeLocation::FromStringView(sql);
   std::vector<ParseToken> parse_tokens;
   std::vector<ParseToken> keywords;
 
   absl::Status status =
       GetParseTokens(ParseTokenOptions(), &location, &parse_tokens);
 
-  if (!status.ok()) return keywords;
+  if (!status.ok()) {
+    std::cout << "Tokenizer Failed for check [" << code
```
Comment (a-litvinov) [inline src/checks_util.cc:301]: I'd rephrase it other way round. The main message here is that we are skipping a check.
`std::cout << "Skipping check [" << code << "] due to tokenizer error: " << status.message()`

File: src/checks_util.h [hunk 36]
```diff
@@ -81,13 +81,17 @@ class ASTNodeRule {
       rule_;
 };
 
+// It will return the corresponding string of a range in a sql.
```
Comment (a-litvinov) [inline src/checks_util.h:84]: Returns the sql substring corresponding to the given range.

File: src/checks_util.h [hunk 37]
```diff
@@ -81,13 +81,17 @@ class ASTNodeRule {
...
 absl::string_view GetName(const ParseLocationRange &range,
                           const absl::string_view &sql);
 
 // Given an ASTNode returns corresponding string for that node.
 absl::string_view GetNodeString(const ASTNode *node,
                                 const absl::string_view &sql);
 
+// Return start position of an ASTNode.
```
Comment (a-litvinov) [inline src/checks_util.h:92]: Returns

File: src/checks_util.cc (hunk 33 above)
Comment (a-litvinov) [reply src/checks_util.cc:40]: Ahm, does it mean that you don't need this function anymore? :) If it is not used anywhere - just inline it into GetNodeString

**Review by a-litvinov — APPROVED** [review]
Please fix the remaining nits and you can merge this PR in.

File: src/checks_util.h (hunk 37 above)
Comment (orhanuysal-google) [reply src/checks_util.h]: fixed

File: src/checks_util.h (hunk 36 above)
Comment (orhanuysal-google) [reply src/checks_util.h]: Erase GetName

File: src/checks_util.cc (hunk 35 above)
Comment (orhanuysal-google) [reply src/checks_util.cc]: Fixed

File: src/checks.cc [hunk 38]
```diff
@@ -68,36 +68,19 @@ LinterResult CheckLineLength(absl::string_view sql,
...
-      location++;
-    if (location >= sql.size() || sql[location] != ';') {
-      if (options.IsActive(ErrorCode::kSemicolon, location))
-        result.Add(ErrorCode::kSemicolon, sql, location,
-                   "Each statement should end with a "
-                   "semicolon ';'.");
-    }
-  }
+  int location = static_cast<int>(sql.size()) - 1;
+  if (!IgnoreSpacesBackward(sql, &location) && sql[location] != ';')
```
[the remark this answers is not in the document]
Comment (orhanuysal-google) [reply src/checks.cc]: Yes, I did a quick fix, and add tests to make sure it doesn't happen again.

File: src/checks_util.cc (hunk 33 above)
Comment (orhanuysal-google) [reply src/checks_util.cc]: Yes, I erased "GetName"
